### PR TITLE
desitorrents: migrate to c#. resolves #11904

### LINKED ISSUE #11904: [desitorrents] Found no results while trying to browse this tracker (Test)
### Do not delete the template.

**Please use the search bar** at the top of the page and make sure you are not creating an already submitted issue.
Check closed issues as well, because your issue may have already been fixed. Also check our [Troubleshooting](https://github.com/Jackett/Jackett/wiki/Troubleshooting) for steps to resolve common issues.

Please read our [Contributing Guidelines](https://github.com/Jackett/Jackett/blob/master/CONTRIBUTING.md) before submitting your issue to ensure a prompt response to your bug.

### Environment

**OS**:

**.Net Runtime**: [.Net-Core/.Net-Framework/Mono]

**.Net Version**:

**Jackett Version**:

**Last Working Jackett Version**:

**Are you using a proxy or VPN?** [yes/no]

### Description

They recently moved their site to Gazelle and updated their API's, now when I connect and test it says "Found no results while trying to browse this tracker". Let me know if you need anything from the dev team and I can get the information as needed.

### Logged Error Messages

[Place any relevant error messages you noticed from the logs here.]

[Make sure you attach the full logs with your personal information removed in case we need more information]

### Screenshots

[Place any screenshots of the issue here if needed]

### PR DESCRIPTION
**Pull request by ilike2burnthing**

### CODE REVIEW

**Comment by ilike2burnthing** [comment]
bad bot...

**Comment by ilike2burnthing** [comment]
Logs in fine, search returns feed, but it errors out with:
```
Error on 20210613021003 for DesiTorrents.txtSystem.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString)
   at Jackett.Common.Indexers.Abstract.GazelleTracker.PerformQuery(TorznabQuery query) in D:\a\1\s\src\Jackett.Common\Indexers\Abstract\GazelleTracker.cs:line 232
```
```
Jackett.Common.IndexerException: Exception (desitorrents): Invalid URI: The format of the URI could not be determined.
 ---> System.Exception: Parse error
 ---> System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString)
   at Jackett.Common.Indexers.Abstract.GazelleTracker.PerformQuery(TorznabQuery query) in D:\a\1\s\src\Jackett.Common\Indexers\Abstract\GazelleTracker.cs:line 232
   --- End of inner exception stack trace ---
   at Jackett.Common.Indexers.BaseWebIndexer.OnParseError(String results, Exception ex) in D:\a\1\s\src\Jackett.Common\Indexers\BaseIndexer.cs:line 761
   at Jackett.Common.Indexers.Abstract.GazelleTracker.PerformQuery(TorznabQuery query) in D:\a\1\s\src\Jackett.Common\Indexers\Abstract\GazelleTracker.cs:line 265
   at Jackett.Common.Indexers.DesiTorrents.PerformQuery(TorznabQuery query) in D:\a\1\s\src\Jackett.Common\Indexers\DesiTorrents.cs:line 62
   at Jackett.Common.Indexers.BaseIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in D:\a\1\s\src\Jackett.Common\Indexers\BaseIndexer.cs:line 385
   --- End of inner exception stack trace ---
   at Jackett.Common.Indexers.BaseIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in D:\a\1\s\src\Jackett.Common\Indexers\BaseIndexer.cs:line 393
   at Jackett.Common.Indexers.BaseWebIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in D:\a\1\s\src\Jackett.Common\Indexers\BaseIndexer.cs:line 735
   at Jackett.Common.Services.IndexerManagerService.TestIndexer(String name) in D:\a\1\s\src\Jackett.Common\Services\IndexerManagerService.cs:line 307
   at Jackett.Server.Controllers.IndexerApiController.Test() in D:\a\1\s\src\Jackett.Server\Controllers\IndexerApiController.cs:line 132
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Jackett.Server.Middleware.CustomExceptionHandler.Invoke(HttpContext httpContext) in D:\a\1\s\src\Jackett.Server\Middleware\CustomExceptionHandler.cs:line 61
```
Which points to:
https://github.com/Jackett/Jackett/blob/7a28412d22cbb5fd58ec204dc36d30fdc552445c/src/Jackett.Common/Indexers/Abstract/GazelleTracker.cs#L230-L232
External posters are formatted as `"cover":"https:\/\/img.onaroll.net\/images\/75abe608de9ee2bfa5f7c1e5260a5b0e.md.jpg"` whereas local posters are `"cover":"2021\/original_152222_manual_3.jpg"` (actual URL is `https://desitorrents.tv/static//media/posters/2021/original_152222_manual_3.jpg`). I think it's tripping over the latter, though I have no idea how to resolve that.

**Comment by ilike2burnthing** [comment]
@seeyabye does the above make sense as cause of this issue?

If so, @LokiThor2021 is this resolvable on your side?

**Comment by ilike2burnthing** [comment]
Just realising `ajax.php` is missing `category` as well.

`CategoryID` is returned in `torrents.php?action=json` but much of it isn't compatible with Gazelle, so would require an actual C# dev to make a separate indexer for it.

**Comment by seeyabye** [comment]
> @seeyabye does the above make sense as a cause of this issue?
> 
> If so, @LokiThor2021 is this resolvable on your side?

The error is definitely caused by a badly formatted poster URL string that was then passed into `Uri()`. It would require additional checks in `GazelleTracker`'s abstract for badly formatted poster Url.

> Just realising `ajax.php` is missing `category` as well.
> 
> `CategoryID` is returned in `torrents.php?action=json` but much of it isn't compatible with Gazelle, so would require an actual C# dev to make a separate indexer for it.

I don't have an account for this tracker, so I can't help much, but if someone has an invite, I could try writing the indexer.

**Comment by LokiThor2021** [comment]
Posters are messed up and is something i plan to fix soon regarding the category in ajax.php what's the endpoint for that ?

**Comment by LokiThor2021** [comment]
@seeyabye  can you share your email i can invite you :)

**Comment by ilike2burnthing** [comment]
Category is just `category` (like grabs is `snatches` and freelech is `isFreeleech`) and is the category ID number:
https://github.com/Jackett/Jackett/blob/313a2fd7845fec1cb853094436a4592841a3b60d/src/Jackett.Common/Indexers/Abstract/GazelleTracker.cs#L345-L357

**Comment by LokiThor2021** [comment]
added category to  ajax.php?action=browse

**Comment by ilike2burnthing** [comment]
Thanks, it needs to be the category ID number though, so `1` for movies, `2` for tv shows, etc.

Not sure if it's case sensitive, but it's a lower case `c` in `category`.

**Comment by seeyabye** [comment]
@LokiThor2021 please send it to seeyabye91 [at] gmail.com. Thanks.

**Comment by LokiThor2021** [comment]
@ilike2burnthing fixed

**Comment by geek-baba** [comment]
Looks like issue is resolved but not able to merge this with master?

**Comment by seeyabye** [comment]
@ilike2burnthing 

I took a look at this, and it's a little bit more complicated given that it's not quite compatible with `GazelleTracker.cs` right out of the box. Several issues that requires fixing:

- Posters are occasionally returning relative links. A simple fix on `GazelleTracker.cs` @ line 231 is to include an additional check
```c#
if (!string.IsNullOrEmpty(cover) && cover.StartsWith("http"))
```
- The categories are not showing if a filter is applied. A temporary fix would be to assume that categories are filtered correctly. See [Oppaitime.cs](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/OppaiTime.csl).

The last issue will require some refactoring work to be done in `GazelleTracker.cs`. When I can find the time, I will work on that, but for now, `GazelleTracker.cs` isn't quite abstract enough to fit all scenarios.

**Comment by ilike2burnthing** [comment]
@seeyabye 
> if (!string.IsNullOrEmpty(cover) && cover.StartsWith("http"))

This worked, thanks.

> The categories are not showing if a filter is applied. A temporary fix would be to assume that categories are filtered correctly.

Do you mean when searching by category, categories aren't shown? I had no issue with that, `https://desitorrents.tv/ajax.php?action=browse&order_by=time&order_way=desc`
and `https://desitorrents.tv/ajax.php?action=browse&order_by=time&order_way=desc&filter_cat[1]=1`
both show `category`.

Rather the issue I'm running into at the moment is that in Jackett all categories are being displayed as Movies, despite them being correctly labelled in the json output.

**Comment by seeyabye** [comment]
> Do you mean when searching by category, categories aren't shown? I had no issue with that, `https://desitorrents.tv/ajax.php?action=browse&order_by=time&order_way=desc`
> and `https://desitorrents.tv/ajax.php?action=browse&order_by=time&order_way=desc&filter_cat[1]=1`
> both show `category`.
> 
> Rather the issue I'm running into at the moment is that in Jackett all categories are being displayed as Movies, despite them being correctly labelled in the json output.

It's probably a related issue. `GazelleTracker` currently defaults all categories that it can't map to `10001` custom category. You will see categories for movies only like `10001`. Filtering by category with 'Tv Shows' which result in `10002` and above will return no results.

**Comment by ilike2burnthing** [comment]
Yea, saw that as well, I assumed it was because they were all being classed as Movies. Given that `category` is present and showing the correct categories, why isn't Jackett using it?

**Comment by seeyabye** [comment]
The category returned by desitorrents is under `group` while `GazelleTracker.cs` expects it to be under `torrent`

<details><summary>sample json result</summary>

```json
{
  "status": "success",
  "response": {
    "currentPage": 1,
    "pages": 77,
    "results": [
      {
        "groupId": 152853,
        "groupName": "Khwabon Ke Parindey",
        "cover": "https://www.socialketchup.in/wp-content/uploads/2021/06/IMG-20210604-WA0003.jpg",
        "tags": [
          "drama",
          "comedy",
          "jatt"
        ],
        "bookmarked": false,
        "groupYear": 0,
        "groupTime": 1623661783,
        "maxSize": 2125630200,
        "totalSnatched": 1,
        "totalSeeders": 2,
        "totalLeechers": 0,
        "category": 2,
        "torrents": [
          {
            "torrentId": 199397,
            "media": "Web",
            "container": "MKV",
            "codec": "x265",
            "resolution": "1080p",
            "audio": "AC3 2.0",
            "lang": "Hindi",
            "subbing": "",
            "archive": null,
            "fileCount": 6,
            "time": "2021-06-14 11:09:43",
            "size": 2125630200,
            "snatches": 1,
            "seeders": 2,
            "leechers": 0,
            "isFreeleech": true,
            "isNeutralLeech": false,
            "isPersonalFreeleech": false,
            "canUseToken": false,
            "hasSnatched": false
          }
        ]
      },
```

</details>

Perhaps @LokiThor2021 could consider changing this part, otherwise, it will have to wait until I figure out how I could refactor out `GazelleTracker.cs`.

**Comment by LokiThor2021** [comment]
@seeyabye will fix it soon

File: src/Jackett.Common/Indexers/DesiTorrents.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuery query)
+        {
+            var releases = await base.PerformQuery(query);
+            foreach (var release in releases)
+            {
+                release.MinimumRatio = 0.6;
+                release.MinimumSeedTime = 259200;
+            }
+            return releases;
+        }
```
Comment (seeyabye) [inline src/Jackett.Common/Indexers/DesiTorrents.cs:69]: Add the following additional function and it should fix the issue on categories:

```c#
        protected override bool ReleaseInfoPostParse(ReleaseInfo release, JObject torrent, JObject result)
        {
            // Add missing category information
            var category = (string)result["category"];
            release.Category = MapTrackerCatToNewznab(category);
            return true;
        }
```

**Comment by seeyabye** [comment]
@LokiThor2021 my apologies, but perhaps there's no need to change on your end. Technically the returned results by desitorrents are correct.

@ilike2burnthing I've proposed a small change and it should fix the issue.

**Review by seeyabye — CHANGES_REQUESTED** [review]

**Comment by ilike2burnthing** [comment]
Yep, that's working fine. Thanks to both of you for your help.

**Pull request merged**